pkg/resources/helpers.go

```diff
@@ -142,3 +146,77 @@ func IsDataType() schema.SchemaValidateFunc { //nolint:staticcheck
 		return warnings, errors
 	}
 }
+
+// IsValidIdentifier is a validator that can be used for validating identifiers passed in resources and data sources.
+// Typically, we expect passed identifiers to be a variation of sdk.ObjectIdentifier. To use this function, pass it as
+// a validation function on identifier field with generic parameter set to the desired sdk.ObjectIdentifier implementation.
+func IsValidIdentifier[T sdk.ObjectIdentifier]() schema.SchemaValidateDiagFunc {
+	return func(value any, path cty.Path) diag.Diagnostics {
+		var diags diag.Diagnostics
+
+		// For now, we won't support sdk.ExternalObjectIdentifiers. The reason behind it is that the functions that parse identifiers are not
+		// able to differentiate between sdk.ExternalObjectIdentifiers and sdk.DatabaseObjectIdentifier or sdk.SchemaObjectIdentifier,
+		// because sdk.ExternalObjectIdentifiers has varying parts count (2 or 3).
+		if _, ok := any(sdk.ExternalObjectIdentifier{}).(T); ok {
+			diags = append(diags, diag.Diagnostic{
+				Severity:      diag.Error,
+				Summary:       "Invalid schema identifier type",
+				Detail:        "Identifier validation is not available for sdk.ExternalObjectIdentifier type. This is a provider error please file a report: https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/new/choose",
+				AttributePath: path,
+			})
+			return diags
+		}
+
+		if stringValue, ok := value.(string); ok {
```

Comment:
nit-pick:
It would be a bit more readable to inverse the check here (if not string return an error - fail fast; then we won't have the nesting here - it would look more go like - the sequence of checking different errs basically).

Reply:
Yeah, I was thinking about it, but ultimately left it as it is. I'll change it.